README.md
# FireMap

[![npm](https://img.shields.io/npm/l/firemap.svg)](https://github.com/artisin/firemap/blob/master/LICENSE.txt)
[![npm](https://img.shields.io/npm/v/firemap.svg)](https://www.npmjs.com/package/firemap)
[![David](https://img.shields.io/david/artisin/firemap.svg)](https://github.com/artisin/firemap/blob/master/package.json)

FireMap is a refinement on the typical "heatmap" to help better visualize mouse position data through Inverse Distance Weighting. I created a companion website to display various FireMap example outputs along with some additional detail you can check out here: [https://firemap.netlify.com/](https://firemap.netlify.com/)

Compared to [heatmap.js](https://www.patrick-wied.at/static/heatmapjs/?utm_source=gh) and other heatmap libraries there are two major differences. FireMap uses a standardized GIS algorithm compared to heatmap.js and others that use a density equation of some sort? This was the primary reason for creating FireMap since I thought it would be advantageous to employ a deterministic algorithm. Secondly, due to the calculation intensive nature of FireMap it's not intended to be used in real-time, although, it does have said capability.

# Install

You can install FireMap either through npm:

```bash
   npm install --save-dev firemap
```

Alternatively, you can manually download FireMap through the Github repository here: [github.com/artisin/firemap/dist/](https://github.com/artisin/firemap/dist)


# Initialize

The FireMap class instance is exported through: `export default`;

__ECMAScript 6 Module Syntax__
```js
import FireMap from 'firemap';
```

__CommonJS Syntax__
```js
const FireMap = require('firemap').default;
```

__Create a New Instance__

FireMap is a Class, and as such, you must initialize it. The FireMap `constructor` accepts an option object argument to set the initial/default options.

```js
// option-less
const firemap = new FireMap();

// options
const firemap = new FireMap({
  // if no canvas element is passed it defauls to the
  // first getElementsByTagName('canvas')
  canvas: document.getElementById('my-canvas-element'),
  // changes sampling DOM area
  area: 10
});
```

## Data Structure

Typically you'll use FireMap in a post-event fashion using tracking data you've collected. However, you can draw a semi-real-time heatmap using the `realTime` method. There are two data types, `dataPoints` (mousemove) and `clickPoints` (pointerdown) that are either passed to the initial Class `constructor` or the `draw` method.

__`dataPoints` → Mouse Position Data__

```js
firemap.draw({
  dataPoints: [{ x: 10, y: 15, value: 5}, ...]
})
```

__`clickPoints` → Mouse Click Data__

```js
firemap.draw({
  clickPoints: [{ x: 20, y: 150}, ...]
})
```

__Both__

```js
firemap.draw({
  dataPoints: [{ x: 10, y: 15, value: 5}, ...],
  clickPoints: [{ x: 20, y: 150}, ...]
})
```


## Class Methods

### Class Instance → `constructor(options = {})`

Creates a `FireMap` instance — many of these options can also be passed to the `draw` method.

+ `options.dataPoints`  = {arr} → user data points for mouse (mousemove) 
+ `options.clickPoints` = {arr} → user data points for clicks (pointerdown)
+ `options.area`        = {num} → sampling area that clusters all points within its area into a single cluster default is 10 so 10px by 10px sample area
+ `options.canvas`      = {dom} → canvas dom element to draw map on
+ `options.cleanEdges`  = {bln} → cleans edges of polygon to remove rough edges to make it look pretty, only used if corners is false
+ `options.clickColor`  = {str} → color of the click point data stars
+ `options.clickSize`   = {num} → size of the click point data stars
+ `options.corners`    = {num}  → creates pseudo data points for corner of the window so heatmap spans the entire screen
+ `options.height`      = {num} → height of canvas || screen height
+ `options.hue`         = {num} → color hue for map default is 0.5 green, 0.8 violet, 3.5 rgbiv and no non-point color
+ `options.interval`    = {num} → interpolation interval of unknown points, the lower the number the more points calculated - computation time increase by O(n2) where n is the number of data points
+ `options.limit`       = {num} → search neighborhood limit, higher number the smoother blend of map colors - has a minor increase in computation time
+ `options.maxValue`    = {num} → used for color, and the default is 100, so any value above 100 is going to be the same color
+ `options.mesh`        = {bln} → to create a mesh-like map rather then a solid map
+ `options.opacity`     = {num} → opacity of canvus fill
+ `options.points`      = {bln} → to draw data marker point coordinates
+ `options.pointSize`   = {num} → font-size of points in px
+ `options.styles`      = {obj} → custom CSS canvas styles
+ `options.subscribe`   = {fnc} → a subscribe function that is invoked on the mouse tracking & click event, passes the event and binded to this
+ `options.threshold`   = {num} → point label value threshold, if a values does not meet threshold no point label will be generated
+ `options.throttle`    = {num} → mouse tracker throttle
+ `options.width`       = {num} → width of canvas, defaults to current screen width

### Draws/create Map → `draw(options = {})`

Draws/creates a heatmap for the canvas element using either passed in `dataPoints` or real-time data points through the invocation of the `draw` method or automatically through the `realTime` method.

__IMPORTANT__: In all likelihood you don’t want data sets of over 1000+ data points because the computation time is O(n2) where n is the number of data points. With 1000-1500 points with an `interval` of `8` it takes 20-40 seconds to compute 1000-1500 data points and if you want a high quality render with an `interval` of `4` there will be two times as many calculations. If you are using FireMap to gather data points you can reduce data points by increasing the `area` (recommended) or increasing the `throttle`.

__Options__
These options can also be passed into the initial `constructor`. 

+ `options.dataPoints` = {arr} → user data points for mouse (mousemove) 
+ `options.clickPoints`= {arr} → user data points for clicks (pointerdown)
+ `options.clickSize`  = {num} → size of the click point data stars
+ `options.clickColor` = {str} → color of the click point data stars
+ `options.cb`         = {fnc} → Callback function invoked upon completion and binded to instance with the first arg being the canvas context (ctx)
+ `options.hue`        = {num} → color hue for map default is 0.5 green, 0.8 violet, 3.5 rgbiv and no non-point color
+ `options.interval`   = {num} → interpolation interval of unknown points, the lower the number the more points calculated - computation time increase by O(n2) where n is the number of data poi
+ `options.limit`      = {num} → search neighborhood limit, higher num smoother blend of map colors
+ `options.mesh`       = {bln} → to create a mesh-like map rather then a solid map
+ `options.opacity`    = {num} → opacity of canvus fill
+ `options.points`     = {bln} → to draw data marker point coordinates
+ `options.pointSize`  = {num} → font-size of points
+ `options.threshold`  = {num} → point values has to be higher than threshold


### Get/Format Data → `getData()`

The `getData` method converts the raw tracking data matrix into valid/formatted tracking data to be used by the `draw` method. However, you'll first need to initialize the tracking feature through the `init` method in order to generate said data to be formatted. The data is returned in an array that consists of objects `{x: <num>, y: <num>, value: <num>}`.


### Initialize Mouse & Click Tracking → `init(subscribe = fnc)`

To use FireMaps built-in mouse position and click logging feature you need to invoke the `init` method. There're two primary ways to "log" this data to send to your server. You can dump the data via the `getData` method either on a leave event or poll event. Or you can subscribe to the raw event data through the `subscribe` function. If you choose the latter, you can pass a `subscribe` function to the `init` method or declare it via the `option` object in the class constructor. The `subscribe` function is binded to the instance so that you can access the internals through `this`. Additionally, the `subscribe` function is passed two arguments, the first is the raw event and the second is the type of event which will be a string of `'mousemove'` or `'pointerdown'`.


### Real Time Drawing → `realTime(drawInterval = 10)`

FireMap can also draw in real-time, but unlike [heatmap.js](https://www.patrick-wied.at/static/heatmapjs/?utm_source=gh), FireMap is not intended to draw in real time. The `drawInterval` determines how often the map should re-draw and the default is set at `10`, so that it re-draws every `10` new data points.



CHANGELOG.md
<a name="0.1.0"></a>
# 0.1.0 (2017-06-15)

* release(0.1.0): Release: 0.1.0 ([2b86869](https://github.com/artisin/firemap/commit/2b86869))
* init(all): init commit → ★ ([c8675ec](https://github.com/artisin/firemap/commit/c8675ec))





__tests__/index.js
require("regenerator-runtime/runtime");
const JSDOM = require('jsdom');
// const cunt = require('canvas-prebuilt')
// const cheerio = require('cheerio')
// const $ = cheerio.load('<canvus id="test"></canvus>')

const dom = new JSDOM.JSDOM(`<body>
  <canvas id="test"></canvas>
  <canvas id="cunt"></canvas>

</body>`);
window = dom.window;
document = window.document;

const fucker = dom.window.document.getElementById('test');
// console.log(fucker.toDataURL())

const FireMap = require('firemap').default;
// import {FireMap} from 'firemap'
// import test from 'firemap';

console.log(FireMap)

// // console.log(fucker.toDataURL())
// const testData = {
//   dataPoints: [{"x":1920,"y":995,"value":11.1},{"x":1920,"y":0,"value":11.1},{"x":1445,"y":180,"value":42.3},{"x":1440,"y":180,"value":10},{"x":1440,"y":175,"value":239},{"x":1375,"y":195,"value":10},{"x":1320,"y":210,"value":10.4},{"x":1300,"y":165,"value":10.1},{"x":1125,"y":360,"value":10.2},{"x":900,"y":135,"value":10.2},{"x":855,"y":125,"value":10},{"x":805,"y":105,"value":10},{"x":785,"y":655,"value":10.1},{"x":670,"y":90,"value":11.7},{"x":660,"y":80,"value":10.2},{"x":655,"y":80,"value":11.7},{"x":650,"y":80,"value":39.1},{"x":645,"y":75,"value":10},{"x":610,"y":380,"value":240.7},{"x":605,"y":545,"value":10.1},{"x":595,"y":765,"value":10},{"x":595,"y":755,"value":10},{"x":595,"y":725,"value":10.2},{"x":590,"y":795,"value":320.8},{"x":590,"y":785,"value":10.7},{"x":590,"y":780,"value":10},{"x":590,"y":775,"value":31.5},{"x":590,"y":770,"value":10.6},{"x":570,"y":385,"value":10.2},{"x":440,"y":410,"value":10},{"x":40,"y":525,"value":10},{"x":0,"y":995,"value":11.1},{"x":0,"y":0,"value":11.1}],
//   clickPoints: [{"x":650,"y":80,"value":1},{"x":590,"y":795,"value":1}]
// };

// // const cunt = FireMap(Object.assign(testData))

// // console.log(fucker.toDataURL())


// // console.log(fucker.toDataURL())
// const cunt = new FireMap(Object.assign({canvas: dom.window.document.getElementById('cunt')}, testData))
// // console.log(fucker.toDataURL())


// cunt.draw({
//   cb: function (ctx) {
//     console.log(dom.window.document.getElementById('cunt').toDataURL())
//     // console.log(ctx.get)
//     // console.log('cunt')
//     // const shit = dom.window.document.getElementById('cunt ');
//     // // console.log(fucker.toDataURL())
//     // console.log(shit.toDataURL())
//     // console.log('dsdsdsdsd')
//     // console.log(this.canvas.toDataURL())
//   }
// });


// // $.html()

// debugger

lib/get-color.js
/**
 * Converts an HSL color value to RGB. Conversion formula
 * adapted from http://en.wikipedia.org/wiki/HSL_color_space.
 * Assumes h, s, and l are contained in the set [0, 1] and
 * returns r, g, and b in the set [0, 255].
 */
const hslToRgb =  function (h, s, l) {
  let r, g, b;

  if(s === 0) {
    // achromatic
    r = g = b = l;
  }else{
    const hue2rgb = function (p, q, t) {
      if(t < 0) { t += 1; }
      if(t > 1) { t -= 1; }
      if(t < 1 / 6) { return p + (q - p) * 6 * t; }
      if(t < 1 / 2) { return q; }
      if(t < 2 / 3) { return p + (q - p) * (2 / 3 - t) * 6; }
      return p;
    };

    const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
    const p = 2 * l - q;
    r = hue2rgb(p, q, h + 1 / 3);
    g = hue2rgb(p, q, h);
    b = hue2rgb(p, q, h - 1 / 3);
  }

  return [Math.round(r * 255), Math.round(g * 255), Math.round(b * 255)];
};

/**
 * Gets color based on constrants
 */
const getColor = function ({hue, max, val}) {
  // 0   -> orange
  // 0.5 -> green
  // 1   -> violet
  const min = 0;
  const dif = max - min;
  val = val > max ? max : val < min ? min : val;
  return hslToRgb(1 - (1 - hue) - (((val - min) * hue) / dif), 1, 0.5);
};

module.exports = getColor;





webpack.config.js
const fs            = require('fs');
const path          = require('path');
const _             = require('lodash');
const webpack       = require('webpack');
const nodeExternals = require('webpack-node-externals');
const dev           = process.env.NODE_ENV === 'dev';

/**
 * Base Webpack Config
 */
const base = {
  entry: {
    'firemap': path.resolve(__dirname, './lib/index.js')
  },
  devtool: dev ? '#eval-source-map' : 'source-map',
  output: {
    path: path.resolve(__dirname, './dist'),
    filename: '[name].js',
    library: '[name]',
    libraryTarget: 'umd'
  },
  module: {
    loaders: [{
      test: /\.js$/,
      exclude: /node_modules/,
      loaders: ['babel-loader']
    }]
  },
  plugins: dev ? [
    new webpack.NoEmitOnErrorsPlugin(),
    new webpack.LoaderOptionsPlugin({debug: true})
  ] : [
    new webpack.optimize.OccurrenceOrderPlugin(true),
    // Appends eslint-disable
    new webpack.BannerPlugin({
      banner: '\n\n/* eslint-disable */\n',
      raw: true
    }),
    // Appends the License into the library builds
    new webpack.BannerPlugin(fs.readFileSync(path.join(process.cwd(), './LICENSE.txt'), 'utf8')),
    new webpack.LoaderOptionsPlugin({
      minimize: true,
      debug: false
    })
  ]
};


/**
 * Node Build
 */
const node = _.defaultsDeep({}, base);
node.target = 'node';
node.externals = [nodeExternals()];

/**
 * Web Build
 */
const web = _.defaultsDeep({}, base);
web.target = 'web';
web.output.filename = '[name].js';

/**
 * Web Min Build
 */
const webMin = _.defaultsDeep({}, base);
webMin.target = 'web';
webMin.output.filename = '[name].min.js';
if (!dev) {
  webMin.plugins.push(new webpack.optimize.UglifyJsPlugin());
}


module.exports = dev
               ? [web]
               :
                 // Remove or comment out if you do not want
                 // webpack to build a specific version
                 // node, web, webMin
                 [webMin, web];



lib/draw-heat-map.js
const getPointValue = require('./get-point-value');
const getColor      = require('./get-color.js');
//requestIdleCallback Polyfill
require('ric');

/**
 * sorts based on i axis
 * @param  {arr} dataPoints -> dataPoints to min/max
 * @param  {str} i          -> x | y
 * @return {obj}            -> min/max
 */
const getMinMax = function (dataPoints, i = 'x') {
  dataPoints.sort(function (a, b) {
    return a[i] === b[i] ? a.x - b.x : a[i] - b[i];
  });
  return {
    min: dataPoints[0][i],
    max: dataPoints[dataPoints.length - 1][i]
  };
};


/**
 * Creates and draws the heat map
 * @param  {ojb} options.ctx        -> canvuas context
 * @param  {num} options.limit      -> neighborhood limit
 * @param  {num} options.interval   -> interpolation interval
 * @param  {arr} options.dataPoints -> data points to create map from [{}...]
 * @param  {arr} options.polygon    -> the surounding convex polygon around dataPoints
 * @param  {bln} options.cleanEdges -> to remove rough edges
 * @param  {bln} options.mesh       -> to create mesh-like map
 * @param  {bln} options.points     -> to draw coordinates for dataPoints
 * @param  {num} options.pointSize  -> size of coordinates points
 */
const drawHeatMap = function ({
  ctx, limit, interval, dataPoints, polygon, cleanEdges, mesh, points, pointSize
}) {
  const self    = this;
  const size    = self.size;
  const hue     = self.hue;
  const max     = self.maxValue;
  const opacity = self.opacity;
  const thresh  = self.threshold;
  const corners = self.corners;
  const radius  = mesh ? interval * 1.5 : 1;
  return new Promise(function (done) {
    window.requestUserIdle(function () {
      //set up
      limit = limit > dataPoints.length ? dataPoints.length : limit + 1;
      const xMinMax = getMinMax(dataPoints, 'x');
      const yMinMax = getMinMax(dataPoints, 'y');
      const xMin = xMinMax.min;
      const yMin = yMinMax.min;
      const xMax = xMinMax.max;
      const yMax = yMinMax.max;
      const dbl = 2 * interval;
      let col = [];
      let val = 0.0;
      let x = 0;
      let y = 0;
      let color = '';
      let gradient = null;
      //clear pervious canvus before redraw
      ctx.clearRect(0, 0, size.width, size.height);

      //draw heatmap
      for (x = xMin; x < xMax; x += interval) {
        for (y = yMin; y < yMax; y += interval) {
          //get indv points value
          val = getPointValue(limit, polygon, dataPoints, {x, y});
          if (val !== -255) {
            ctx.beginPath();
            //get corresponding color to val
            col = getColor({val, hue, max});
            color = `rgba(${col[0]}, ${col[1]}, ${col[2]},`;
            gradient = ctx.createRadialGradient(x, y, radius, x, y, interval);
            gradient.addColorStop(0, color + opacity + ')');
            gradient.addColorStop(1, color + ' 0)');
            ctx.fillStyle = '#ffffff';
            ctx.lineJoin = 'round';
            ctx.fillStyle = gradient;
            ctx.fillRect(x - interval, y - interval, dbl, dbl);
            ctx.fill();
            ctx.closePath();
          }
        }
      }


      /**
       * Clean hard edges
       */
      if (!corners && (cleanEdges && polygon.length > 1)) {
        ctx.globalCompositeOperation = 'destination-in';
        ctx.fillStyle = '#ffffff';
        ctx.lineJoin = 'round';
        ctx.beginPath();
        ctx.moveTo(polygon[0].x, polygon[0].y);
        for (let i = 1; i < polygon.length; i++) {
          ctx.lineTo(polygon[i].x, polygon[i].y);
        }
        ctx.lineTo(polygon[0].x, polygon[0].y);
        ctx.closePath();
        ctx.fill();
      }


      /**
       * Draw coordinate points
       */
      if (points) {
        const PI2 = 2 * Math.PI;
        let point = null;
        for (let i = 0; i < dataPoints.length; i += 1) {
          point = dataPoints[i];
          if (point.value > thresh) {
            //get corresponding color to val
            color = getColor({val: point.value, hue, max});
            ctx.globalCompositeOperation = 'source-over';
            ctx.fillStyle = 'rgba(255, 255, 255, 0.95)';
            ctx.beginPath();
            ctx.arc(point.x, point.y, pointSize, 0, PI2, false);
            ctx.fill();
            ctx.lineWidth = pointSize / 4;
            ctx.strokeStyle = `rgba(${color[0]}, ${color[1]}, ${color[2]}, 0.8)`;
            ctx.beginPath();
            ctx.arc(point.x, point.y, pointSize, 0, PI2, false);
            ctx.stroke();
            ctx.textAlign = 'center';
            ctx.font = `${pointSize - 2}px monospace`;
            ctx.textBaseline = 'middle';
            ctx.fillStyle = '#474f50';
            ctx.fillText(Math.round(point.value), point.x, point.y);
          }
        }
      }

      //->
      done(true);
    }, {timeout: 2000});
  });

};

module.exports = drawHeatMap;


lib/index.js
const lodashThrottle       = require('lodash.throttle');
const detectIt             = require('detect-it');
const getConvexHullPolygon = require('./get-convex-hull-polygon.js');
const drawHeatMap          = require('./draw-heat-map.js');
const drawClickMap         = require('./draw-click-map.js');


class FireMap {
  /**
   * @param  {arr} dataPoints       -> user data points for mouse
   * @param  {arr} clickPoints      -> user data points for clicks
   * @param  {num} area             -> sampling area which clusters all points
   *                                   within its area into a single cluster
   *                                   default is 5 so 5px by 5px sample area
   * @param  {dom} canvas           -> canvas dom element to draw map on
   * @param  {bln} cleanEdges       -> cleans edges of polygon to remove rough
   *                                   edges to make it look pretty, only used
   *                                   if corners is false
   * @param  {str} clickColor       -> color of the click point data stars
   * @param  {num} clickSize        -> size of the click point data stars
   * @param  {num}  corners         -> creats pseudo data points for cornor of the
   *                                   window so heatmap spans the entire screen
   * @param  {num} height           -> height of canvas || screen height
   * @param  {num} hue              -> color hue for map default is 0.5 green,
   *                                   0.8 violet, 3.5 rgbiv and no non-point color
   * @param  {num} interval         -> interpolation interval of unknown points,
   *                                   the lower the number the more points
   *                                   calculates which producesed a better map
   * @param  {num} limit            -> search neighborhood limit, higher num
   *                                   smoother blend of map colors
   *                                   but it also increase computation time
   * @param  {num} maxValue         -> max value for color default is 100
   *                                   so any value above 100 is going to be the
   *                                   same color
   * @param  {bln} mesh             -> to create a mesh-like map rather then
   *                                   a solid map
   * @param  {num} opacity          -> opacity of canvus fill
   * @param  {bln} points           -> to draw data marker point coordinates
   * @param  {num} pointSize        -> font-size of points
   * @param  {obj} styles           -> custom CSS canvas styles
   * @param  {fnk} subscribe        -> a subscribe function that is invoked on
   *                                   the mouse tracking & click event,
   *                                   passes the event and binded to this
   * @param  {num} threshold        -> point values has to be higher than threshold
   * @param  {num} throttle         -> mouse tracker throttle
   * @param  {num} width            -> width of canvas || screen width
   */
  constructor({
    //user data points
    dataPoints = [], clickPoints = [], subscribe,
    //cavus setting
    canvas, width, height, area = 10, maxValue = 100, styles = {},
    //map appearance settings
    hue = 0.5, opacity = 0.8, cleanEdges = true,
    //mouse event setting
    throttle = 100, threshold = 110,
    //draw setting
    corners = true, limit = 100, interval = 8, mesh = false, points = false, pointSize = 13,
    //click setting
    clickTrack = false, clickSize, clickColor
  } = {}) {
    const self = this;
    self.error = false;

    /**
     * Error gate
     */
    canvas = canvas
             || document.getElementsByTagName('canvas')
             && document.getElementsByTagName('canvas')[0];
    if (!canvas) {
      self.error = true;
      console.error('Fatal Fire Map Error: Canvas element not found, cannot proceed.');
      return false;
    }
    /**
     * User data
     */
    self.dataPoints  = dataPoints;
    self.clickPoints = clickPoints;

    /**
     * Tracking Vars/setup
     */
    self.area = area;
    self.throttle = throttle;
    self.threshold = threshold;
    const getDocSize = function(axis = 'Width') {
      const D = document;
      return Math.max(
        D.body[`scroll${axis}`], D.documentElement[`scroll${axis}`],
        D.body[`offset${axis}`], D.documentElement[`offset${axis}`],
        D.body[`client${axis}`], D.documentElement[`client${axis}`]
      );
    };
    self.width = width ? width : getDocSize('Width');
    self.height = height ? height : getDocSize('Height');
    //used in mouse track
    self._time = false;
    self._temp = [];
    //create location matrix, add one for safty
    self.data = [];
    self.data.length = Math.round((self.width + self.area) / self.area) + 1;
    const ySize = Math.round((self.height + self.area) / self.area) + 1;
    for (let i = self.data.length - 1; i >= 0; i--) {
      self.data[i] = [];
      self.data[i].length = ySize;
    }
    //checks for clickTracking and creates need be
    self.clickTrack = clickTrack;
    if (self.clickTrack) {
      self.clickData = [];
      self.clickData.length = Math.round((self.width + self.area) / self.area) + 1;
      for (let i = self.clickData.length - 1; i >= 0; i--) {
        self.clickData[i] = [];
        self.clickData[i].length = ySize;
      }
    }

    /**
     * HeatMap
     */
    self.drawHeatMap = drawHeatMap.bind(self);
    self.drawClickMap = drawClickMap.bind(self);
    self.maxValue = maxValue;
    self.hue = hue;
    self.cleanEdges = cleanEdges;
    self.opacity = opacity;
    self.ctx = null;
    //set canvus to take up entire screen
    canvas.width = self.width;
    canvas.height = self.height;
    self.canvas = canvas;
    self.points = [];
    self.polygon = [];
    self.limits = {
      xMin: 0,
      xMax: 0,
      yMin: 0,
      yMax: 0
    };
    self.size = {
      height: canvas.height,
      width: canvas.width
    };
    //set canvus style props
    styles = Object.assign({
      position: 'absolute',
      top: '0',
      left: '0',
      opacity: opacity
    }, styles);
    Object.keys(styles).forEach(function (key) {
      canvas.style[key] = styles[key];
    });

    /**
     * Draw
     */
    self.corners   = corners;
    self.limit     = limit;
    self.interval  = interval;
    self.mesh      = mesh;
    self.points    = points;
    self.pointSize = pointSize;
    //clicks
    self.clickSize = clickSize;
    self.clickColor = clickColor;

    //used for realtime tracking
    if (typeof subscribe === 'function') {
      self.subscribe = subscribe.bind(self);
    }
    self.dataPointCount = false;
  }



  /**
   * Draws the heatzmap based on data points
   * @param  {arr} options.dataPoints -> user data points for mouse
   * @param  {arr} options.clickPoint -> user data points for clicks
   * @param  {num} options.clickSize  -> size of the click point data stars
   * @param  {str} options.clickColor -> color of the click point data stars
   * @param  {fnk} options.cb         -> Callback function invoked upon completion
   *                                     and binded to instance with the first arg
   *                                     being the canvas context (ctx)
   * @param  {num} options.hue        -> color hue for map default is 0.5 green,
   *                                     0.8 violet, 3.5 rgbiv and no non-point color
   * @param  {num} options. interval  -> interpolation interval of unknown points,
   *                                     the lower the number the more points
   *                                     calculates which producesed a better
   * @param  {num} options.limit      -> search neighborhood limit, higher num
   *                                     smoother blend of map colors
   * @param  {bln} options.mesh       -> to create a mesh-like map rather then
   *                                    a solid map
   * @param  {num} options.opacity    -> opacity of canvus fill
   * @param  {bln} options.points     -> to draw data marker point coordinates
   * @param  {num} options.pointSize  -> font-size of points
   * @param  {num} options.threshold  -> point values has to be higher than threshold
   */
  draw ({
    dataPoints, clickPoints, limit, interval,
    mesh, points, pointSize, threshold, hue, opacity, clickSize, clickColor,
    cb = false
  } = {}) {
    const self = this;
    dataPoints  = dataPoints  !== undefined ? dataPoints : self.dataPoints;
    clickPoints = clickPoints !== undefined ? clickPoints : self.clickPoints;
    limit       = limit       !== undefined ? limit : self.limit;
    interval    = interval    !== undefined ? interval : self.interval;
    mesh        = mesh        !== undefined ? mesh : self.mesh;
    points      = points      !== undefined ? points : self.points;
    pointSize   = pointSize   !== undefined ? pointSize : self.pointSize;
    //self vars
    self.threshold  = threshold  !== undefined ? threshold : self.threshold;
    self.hue        = hue        !== undefined ? hue : self.hue;
    self.opacity    = opacity    !== undefined ? opacity : self.opacity;
    self.clickSize  = clickSize  !== undefined ? clickSize : self.clickSize;
    self.clickColor = clickColor !== undefined ? clickColor : self.clickColor;
    let ctx;
    /**
     * Wrapper so that we don't trip up the main thread. draw-heat-map also
     * is wrapped via requestUserIdle to only execue on idle
     */
    const run = async function run () {
      //cannot't clean with real time
      const cleanEdges = dataPoints ? self.cleanEdges : false;
      ctx        = self.canvas.getContext('2d');
      dataPoints       = dataPoints.length ? dataPoints : await self.getData();
      if (dataPoints.length) {
        const polygon    = await getConvexHullPolygon(dataPoints);
        await self.drawHeatMap({
          ctx, limit, interval, dataPoints, polygon, cleanEdges, mesh, points, pointSize
        });
      }
      //clickpoints
      clickPoints = clickPoints === true
                  ? await self.getData(self.clickData || [], true)
                  : clickPoints;
      if (clickPoints.length) {
        await self.drawClickMap({ctx, clickPoints});
      }
      return true;
    };
    run().then(function () {
      console.log('Draw Complete');
      if (cb) {
        cb.apply(self, [ctx]);
      }
    });
  }


  /**
   * Formats the data matricks are removes undefined data
   * @param  {arr} source -> Source data
   * @return {prm}         -> promise([{x, y, value} ...])
   */
  getData (source = null, click = false) {
    const self = this;
    return new Promise(function (done) {
      source = source || self.data;
      const data = [];
      let value  = null;
      for (let x = source.length - 1; x >= 0; x--) {
        for (let y = source[x].length - 1; y >= 0; y--) {
          value = source[x][y];
          if (value) {
            data.push({
              x: x * self.area,
              y: y * self.area,
              value: !click ? value / 10 : value
            });
          }
        }
      }
      done(data);
    });
  }


  /**
   * Client init to start tracking mouse movements
   */
  init(subscribe = null) {
    const self = this;
    if (!self.error) {
      //subscribe set up
      self.subscribe = subscribe && typeof subscribe === 'function'
                     ? subscribe.bind(self)
                     : self.subscribe;
      const sub = (event, type) => self.subscribe && self.subscribe(event, type);
      //mouse movement
      self._trackAndLogMouse = self._trackAndLogMouse.bind(self);
      document.addEventListener('mousemove',
        lodashThrottle(function (event) {
          self._trackAndLogMouse(event);
          sub(event, 'mousemove');
        }, self.throttle),
        detectIt.passiveEvents ? {passive: true} : false
      );

      //click
      if (self.clickTrack) {
        self._logClicks = self._logClicks.bind(self);
        document.addEventListener('pointerdown',
          lodashThrottle(function (event) {
            self._logClicks(event);
            sub(event, 'pointerdown');
          }, 75),
          detectIt.passiveEvents ? {passive: true} : false
        );
      }

      //used so that tracker resets otherwise you get edge outliers
      document.addEventListener('pointerdown', function () {
        self._time = new Date();
      }, detectIt.passiveEvents ? {passive: true} : false);
    }
    //places data points at corners of canvus
    if (self.corners) {
      const val = 1;
      const x = self.data.length - 1;
      const y = self.data[0].length - 1;
      self.data[0][0] = val;
      self.data[0][y] = val;
      self.data[x][0] = val;
      self.data[x][y] = val;
    }
  }


  _setCoordinates(x, y, mouse = true) {
    const self = this;
    const posX = Math.round(x / self.area);
    const posY = Math.round(y / self.area);
    //mouse loging
    if (mouse) {
      if (!self._time || !self._temp.length) {
        self._time = new Date();
        self._temp = [posX, posY];
      }else {
        const newTime = new Date();
        const time = newTime - self._time;
        //set current data
        const temp = self._temp;
        //set data, and error check
        if (self.data[temp[0]]) {
          const data = self.data[temp[0]][temp[1]];
          self.data[temp[0]][temp[1]] = !data ? time : (data + time);
        }
        //store new time + temp
        self._time = newTime;
        self._temp = [posX, posY];
      }
      //for real time update
      if (self.dataPointCount) {
        self.dataPointCount.count += 1;
      }
    }else {
      //click track
      const data = self.clickData[posX][posY];
      self.clickData[posX][posY] = !data ? 1 : (data + 1);
    }
  }


  /**
   * Tracks mouse events set movements to _setCoordinates
   */
  _trackAndLogMouse(event) {
    const self = this;
    event = event || window.event;
    if (event.pageX === null && event.clientX !== null) {
      const eventDoc = (event.target && event.target.ownerDocument) || document;
      const doc      = eventDoc.documentElement;
      const body     = eventDoc.body;

      event.pageX = event.clientX +
                    (doc && doc.scrollLeft || body && body.scrollLeft || 0) -
                    (doc && doc.clientLeft || body && body.clientLeft || 0);
      event.pageY = event.clientY +
                    (doc && doc.scrollTop  || body && body.scrollTop  || 0) -
                    (doc && doc.clientTop  || body && body.clientTop  || 0);
    }
    self._setCoordinates(event.pageX, event.pageY);
  }


  /**
   * Tracks clicks
   */
  _logClicks(event) {
    const self = this;
    event = event || window.event;
    if (event.pageX === null && event.clientX !== null) {
      const eventDoc = (event.target && event.target.ownerDocument) || document;
      const doc      = eventDoc.documentElement;
      const body     = eventDoc.body;

      event.pageX = event.clientX +
                    (doc && doc.scrollLeft || body && body.scrollLeft || 0) -
                    (doc && doc.clientLeft || body && body.clientLeft || 0);
      event.pageY = event.clientY +
                    (doc && doc.scrollTop  || body && body.scrollTop  || 0) -
                    (doc && doc.clientTop  || body && body.clientTop  || 0);
    }
    self._setCoordinates(event.pageX, event.pageY, false);
  }


  /**
   * Updates heatmap in real time
   * @param  {num} interval -> update heatmap every x data points
   */
  realTime (drawInterval = 10) {
    const self = this;
    const dataPointCount = {count: 0};
    const proxy = new Proxy(dataPointCount, {
      set: function (target, property, value) {
        target[property] = value;
        if (value % drawInterval === 0) {
          self.draw();
        }
        return true;
      }
    });
    self.dataPointCount = proxy;
  }
}


export default FireMap;
